libwacom/libwacom-database.c

```diff
@@ -295,32 +419,65 @@ libwacom_parse_stylus_keyfile(WacomDeviceDatabase *db, const char *path)
 
 	groups = g_key_file_get_groups (keyfile, NULL);
 	for (i = 0; groups[i]; i++) {
-		WacomStylus *stylus;
-		WacomStylusId id;
+		WacomStylus *stylus, *other;
+		WacomStylusId id, lookup;
 		GError *error = NULL;
 		char *eraser_type, *type;
 		char **string_list;
+		char *alias_value;
 
 		if (!parse_stylus_id(groups[i], &id)) {
 			g_warning ("Failed to parse stylus ID '%s', ignoring entry", groups[i]);
 			continue;
 		}
+
 		stylus = g_new0 (WacomStylus, 1);
 		stylus->refcnt = 1;
 		stylus->id = id;
-		stylus->name = g_key_file_get_string(keyfile, groups[i], "Name", NULL);
-		stylus->group = g_key_file_get_string(keyfile, groups[i], "Group", NULL);
+
+		if (alias == IGNORE_ALIASES && g_key_file_get_string(keyfile, groups[i], "AliasOf", NULL) != NULL) {
+			continue;
+		}
+
+		if (alias == ONLY_ALIASES && g_key_file_get_string(keyfile, groups[i], "AliasOf", NULL) == NULL) {
+			continue;
+		}
+
+		if (alias == ONLY_ALIASES) {
+			alias_value = g_key_file_get_string(keyfile, groups[i], "AliasOf", NULL);
+
+			if (!create_stylus_id_from_aliased(groups[i], alias_value, &lookup)) {
+				g_warning ("Failed to parse Aliased ID '%s', ignoring entry", groups[i]);
+	                        continue;
+			}
+			other = g_hash_table_lookup(db->stylus_ht, &lookup);
+			if (!other) {
+				g_warning ("Unkown Aliased ID 0x%08x %s\n", lookup.tool_id, alias_value);
+				continue;
+			}
+		}
+
+		stylus->name = string_or_fallback(keyfile, groups[i], "Name", other ? other->name : NULL);
+		stylus->group = string_or_fallback(keyfile, groups[i], "Group", other ? other->group : NULL);
 		stylus->paired_stylus_ids = g_array_new (FALSE, FALSE, sizeof(WacomStylusId));
 
-		eraser_type = g_key_file_get_string(keyfile, groups[i], "EraserType", NULL);
+		eraser_type = string_or_fallback(keyfile, groups[i], "EraserType", other ? eraser_str_from_type(other->eraser_type) : NULL);
 		stylus->eraser_type = eraser_type_from_str (eraser_type);
 		g_clear_pointer(&eraser_type, g_free);
 
 		/* We have to keep the integer array for libwacom_get_supported_styli() */
 		stylus->deprecated_paired_ids = g_array_new (FALSE, FALSE, sizeof(int));
 		stylus->paired_styli = g_array_new (FALSE, FALSE, sizeof(WacomStylus*));
 
-		string_list = g_key_file_get_string_list (keyfile, groups[i], "PairedStylusIds", NULL, NULL);
+		if (alias == IGNORE_ALIASES) {
+			string_list = g_key_file_get_string_list(keyfile, groups[i], "PairedStylusIds", NULL, NULL);
+		} else {
+			string_list = g_key_file_get_string_list(keyfile, groups[i], "PairedStylusIds", NULL, NULL);
```

Comment:
```suggestion
		string_list = g_key_file_get_string_list(keyfile, groups[i], "PairedStylusIds", NULL, NULL);
		if (alias != IGNORE_ALIASES) {
```
does the same thing I think